### ISSUE TITLE
Torch doesn't copy the assigned self-referential memory on `cpu` (inconsistent with `numpy` and `cuda`)

### ISSUE THREAD

**Issue by VladislavZavadskyy** [pytorch/pytorch#126097]
### 🐛 Describe the bug

Consider a simple example:

```python
import torch

a = torch.arange(16).view(4, 4)
a[:3, :3] = a[:3, :3].T

b = torch.arange(16, device='cuda').view(4, 4)
b[:3, :3] = b[:3, :3].T

print(a, b, sep='\n\n')
```
which outputs
```
tensor([[ 0,  4,  8,  3],
        [ 4,  5,  9,  7],
        [ 8,  9, 10, 11],
        [12, 13, 14, 15]])

tensor([[ 0,  4,  8,  3],
        [ 1,  5,  9,  7],
        [ 2,  6, 10, 11],
        [12, 13, 14, 15]], device='cuda:0')
```
Apparently, the CPU version doesn't make a copy of the operand and references the memory directly. If the operand is cloned everything works as expected.

### Versions

PyTorch version: 2.3.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Fedora Linux 40 (Workstation Edition) (x86_64)
GCC version: (GCC) 12.3.0
Clang version: Could not collect
CMake version: version 3.28.2
Libc version: glibc-2.39

Python version: 3.11.9 (main, Apr 17 2024, 00:00:00) [GCC 14.0.1 20240411 (Red Hat 14.0.1-0)] (64-bit runtime)
Python platform: Linux-6.8.9-300.fc40.x86_64-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.4.131
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4080
Nvidia driver version: 550.78
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               32
On-line CPU(s) list:                  0-31
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Core(TM) i9-14900KF
CPU family:                           6
Model:                                183
Thread(s) per core:                   2
Core(s) per socket:                   24
Socket(s):                            1
Stepping:                             1
CPU(s) scaling MHz:                   88%
CPU max MHz:                          6000,0000
CPU min MHz:                          800,0000
BogoMIPS:                             6374,40
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            896 KiB (24 instances)
L1i cache:                            1,3 MiB (24 instances)
L2 cache:                             32 MiB (12 instances)
L3 cache:                             36 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-31
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Vulnerable: No microcode
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.3.0
[pip3] torchaudio==2.3.0
[pip3] torchvision==0.18.0
[pip3] triton==2.3.0

**Comment by VladislavZavadskyy** [comment]
It appears to also be an inconsistency with numpy

```python
a = np.arange(16).reshape(4, 4)
a[:3, :3] = a[:3, :3].T

a
```
outputs
```
array([[ 0,  4,  8,  3],
       [ 1,  5,  9,  7],
       [ 2,  6, 10, 11],
       [12, 13, 14, 15]])
 ```

**Comment by mingfeima** [comment]
@min-jean-cho please help debug this one!

**Comment by ezyang** [comment]
Unfortunately self mutation on overlapping input/output is generally undefined behavior and we don't give guarantees for this case.

**Comment by VladislavZavadskyy** [comment]
It can be a pain to debug or even understand that something is wrong, if someone writes and tests a model entirely on cuda and then someone else runs it on cpu. Surely it can't be impossible to check if the operand references the same memory and make a copy in that case?

**Comment by ezyang** [comment]
Yes, for a case like this it should be possible to detect if overlap has occurred and do something. Maybe erroring is better though :P

**Comment by VladislavZavadskyy** [comment]
I understand, that from the perspective of, say C++, referencing the same memory while it's being copied sequentially is ill-defined behavior (although you can define it for any specific case). But torch is not about that, is it?. It is about manipulating data on scale, if sequential operations happen somewhere under the hood, users doesn't need to burden themselves with thinking about it. This is an abstraction layer, and abstraction layers is what allowed us to get this far, don't think anyone would come up with transformer architecture with only assembly being available :) 
So I believe that numpy-like behavior with copying would be okay, since that what you would expect without thinking how it's implemented on a lower level. But error beats nothing, for sure ;)

**Comment by ngimel** [comment]
We cannot detect this happening without adding considerable overhead to all ops. As @ezyang said it's undefined behavior, we error in some obvious cases that are easy to detect, and generally leave it as a user responsibility to not do self-mutating calls. BTW it is not guaranteed to work on CUDA either, you are getting lucky here because your input is small, but you'll start getting wrong results once the tensor becomes larger.
